source/lac/trilinos_epetra_vector.cc

```diff
@@ -0,0 +1,529 @@
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LinearAlgebra
+{
+  namespace EpetraWrappers
+  {
+    Vector::Vector()
+      :
+      vector(new Epetra_FEVector(Epetra_Map(0,1,0,Epetra_MpiComm(MPI_COMM_SELF))))
+    {}
+
+
+
+    Vector::Vector(const Vector &V)
+      :
+      vector(new Epetra_FEVector(V.trilinos_vector()))
+    {}
+
+
+
+    Vector::Vector(const IndexSet &parallel_partitioner,
+                   const MPI_Comm &communicator)
+      :
+      vector(new Epetra_FEVector(parallel_partitioner.make_trilinos_map(communicator,false)))
+    {}
+
+
+
+    void Vector::reinit(const IndexSet &parallel_partitioner,
+                        const MPI_Comm &communicator,
+                        const bool      omit_zeroing_entries)
+    {
+      Epetra_Map input_map = parallel_partitioner.make_trilinos_map(communicator,false);
+      if (vector->Map().SameAs(input_map)==false)
+        vector.reset(new Epetra_FEVector(input_map));
+      else if (omit_zeroing_entries==false)
+        {
+          const int ierr = vector->PutScalar(0.);
+          (void) ierr;
+          Assert(ierr==0, ExcTrilinosError(ierr));
+        }
+    }
+
+
+
+    Vector &Vector::operator= (const Vector &V)
+    {
+      // Distinguish three cases:
+      //  - First case: both vectors have the same layout.
+      //  - Second case: both vectors have the same size but different layout.
+      //  - Third case: the vectors have different size.
+      if (vector->Map().SameAs(V.trilinos_vector().Map()))
+        *vector = V.trilinos_vector();
+      else
+        {
+          if (size()==V.size())
+            {
+              Epetra_Import data_exchange(vector->Map(), V.trilinos_vector().Map());
+
+              const int ierr = vector->Import(V.trilinos_vector(), data_exchange, Insert);
+              Assert(ierr==0, ExcTrilinosError(ierr));
+            }
+          else
+            vector.reset(new Epetra_FEVector(V.trilinos_vector()));
+        }
+
+      return *this;
+    }
+
+
+
+    Vector &Vector::operator= (const ::dealii::LinearAlgebra::ReadWriteVector<double> &V)
+    {
+      IndexSet elements_to_write(V.get_stored_elements());
+      IndexSet locally_owned_elements(this->locally_owned_elements());
+      IndexSet off_proc_elements(elements_to_write);
+      off_proc_elements.subtract_set(locally_owned_elements);
+      IndexSet on_proc_elements(elements_to_write);
+      on_proc_elements.subtract_set(off_proc_elements);
+
+      // Write the elements that locally owned.
+      IndexSet::ElementIterator elem = on_proc_elements.begin(),
+                                elem_end = on_proc_elements.end();
+      for (; elem!=elem_end; ++elem)
+        {
+          const TrilinosWrappers::types::int_type local_index =
+            vector->Map().LID(static_cast<TrilinosWrappers::types::int_type>(*elem));
+          (*vector)[0][local_index] = V[*elem];
+        }
+
+
+      // Write the elements off-processor. Cannot use Import function of Trilinos
```

Comment:
The problem is that we don't know where the data is. When I use Trilinos, I can use two maps to know on which processor each elements is but that doesn't work with IndexSet. I could create a ReadWriteVector such that every element locally owned needs to be communicated or having ReadWriteVector on only one processor.  If I can create the communication pattern once and then, reuse it then it would be easier to make it scale.